lib/gold_miner/blog_post.rb

```diff
@@ -10,25 +12,30 @@ def initialize(slack_channel:, messages:, since:, writer: SimpleWriter.new)
     end
 
     def to_s
-      <<~MARKDOWN
-        ---
-        title: "#{title}"
-        tags: #{tags}
-        teaser: >
-          Highlights of what happened in our ##{@slack_channel} channel on Slack this week.
-        author: Matheus Richard
-        ---
-
-        Welcome to another edition of This Week in #dev, a series of posts where we
-        bring some of the most interesting Slack conversations to the public.
-        Today we're talking about: #{topics_sentence}.
-
-        #{highlights}
-
-        ## Thanks
-
-        This edition was brought to you by: #{authors}. Thanks to all contributors! 🎉
-      MARKDOWN
+      Sync do
+        tags_task = Async { tags }
+        highlights_task = Async { highlights }
```

Comment:
Just catching up with the `async` gem. I basically only know async via Rust. I think my mental model might be wrong?

If I were to do this in Rust, it'd go like this:

1. `@writer` would have async methods. `SimpleWriter` would use the async-io library; OpenAIWriter would [use the `Async` method](https://socketry.github.io/async/guides/getting-started/index.html#compatibility).
2. `GoldMiner::Slack::Client` methods would also use the `Async` method.
3. Then here, I'd write something like:

```ruby
def to_s
  Sync do |task|
    the_tags = tags.wait
    the_highlights = highlights.wait

    <<~MARKDOWN
    ... #{the_tags} .... #{the_highlights} ....
    MARKDOWN
  end
end

def highlights
  Async do |task|
    @messages
        .map { |message| Async { highlight_from(message) } }
        .map(&:wait)
        .join("\n")
        .chomp("")
  end
end

def highlight_from(message)
  Async do |task|
    title = @writer.give_title_to(message).wait
    summary = @writer.summarize(message).wait

    <<~MARKDOWN
      ## #{title}
      #{summary}
    MARKDOWN
  end
end
```

That is, if you call `#wait`, then you wrap yourself in `Sync` or `Async`.

But maybe that's backward? I think `Kernel#Async` might not do what I expect? Is it used for controlling fibers and flow, and therefore should be used within `Sync` blocks only?

I tried finding good examples and wow there sure aren't any.

Reply:
I honestly can't answer much. I ran into the same problems you're describing.

Aside from the [project docs](https://socketry.github.io/async/), this [blog post](https://brunosutic.com/blog/async-ruby), and [a HackerNews thread](https://news.ycombinator.com/item?id=29049881), I couldn't find much.

I was mostly experimenting on my own to make it work. What you wrote might even work! I'll try it to see how different it is.

Reply:
They keep saying this is like Go, i.e., it has no [function colors](https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/), but I'm not sure yet.

Reply:
OK re-reading the docs now, I think I have an understanding of things.

We have one event loop. We could imagine two ways of modeling things from there:

## two subtasks that run in sequence (containing parallelizable tasks within)

![a top level task for the Slack API and for OpenAI, and then subtasks within](https://user-images.githubusercontent.com/4550/211107152-d5212793-1e4f-4082-9268-7c8f121404cf.png)

This model is kinda like this psuedo-ish code:

```ruby
def run
  Sync do |task|
    messages = Slack::Client.build(whatever).search_interesting_messages_in("#dev", task).wait
    blog_post = BlogPost.new("#dev", messages, whatever).to_text(task).wait
    puts blog_post
  end
end
```

In this model the entire Slack::Client does its thing, returns the messages, then the BlogPost takes over. The `#search_..._in` method and `#to_text` method are quite parallelized within, though:

```ruby
  def search_interesting_messages_in(channel, task)
    task.async do
      barrier = Async::Barrier.new
      interesting_messages = interesting_messages_query(channel)
      til_messages = barrier.async { |task| extract_messages_from_result(
        @slack.search_messages(query: interesting_messages.with_topic("TIL")),
        task,
      ).wait }
      tip_messages = barrier.async { |task| extract_messages_from_result(
        @slack.search_messages(query: interesting_messages.with_topic("tip")),
        task,
      ).wait }
      golden_messages = barrier.async { |task| extract_messages_from_result(
        @slack.search_messages(query: interesting_messages.with_reaction(GOLD_EMOJI)),
        task,
      ).wait }

      barrier.wait

      (til_messages + tip_messages + golden_messages).uniq { |message| message[:permalink] }
    end
  end
```

That uses a barrier to track all the tasks, waiting for them all to end. That might be unnecessary here, I'm still figuring out this tooling.

## all tasks run in parallel

![one event loop with every task as a child](https://user-images.githubusercontent.com/4550/211108504-841fa83a-d33b-4c1c-9db6-6c51355592a4.png)

The other option is to stream data from the Slack API to OpenAI and into stdout. This might have the highest performance, with could also require the greater changes to the code.

```ruby
def run_it_all
  messages = []
  msg_mutex = Mutex.new
  highlights = []
  hi_mutex = Mutex.new
  blog_post = BlogPost.new
  Sync do |task|
    slack = Slack::Client.build.wait
    task.async {|t| slack.search_interesting_messages_in("#dev", t, msg_mutex, messages) }
    task.async {|t| blog_post.highlights(t, msg_mutex, messages, hi_mutx, highlights) }
    # and so on
    task.async {|t| write_out(t, hi_mutex, highlights)
  end
end
```

That needs care because it seems that the only way to stream data is to write to a shared data structure, so you'll need to `mutex.synchronize` around writes/reads on that variable.

Reply:
@mike-burns I appreciate the writing, but I'm still confused. I'm not parallelizing the work on the Slack API yet, just the creation of the blog posts (mainly focusing on the `OpenAiWriter`, which makes several HTTP requests).

I don't think I need to touch anything outside the `BlogPost` class to make it work in parallel. My idea was to make this performance work transparent, so I don't have to rewrite code to things like:

```rb
BlogPost.new("#dev", messages, whatever).to_text(task).wait
```

In other words, the parallelization is "inside" the BlogPost class but not seen by the "outside".